blog/content/post/2020-08-19-system_latency.md

```diff
@@ -0,0 +1,172 @@
+abstract: "What in reality is RT system? This post will explain what to expect from Real-Time systems and how can we test performance in this kind of builds."
+tags:
+  - yocto
+  - testing
+  - latencies
+  - i.mx8
+categories:
+  - Real-Time
+  - Testing
+---
+
+# Latency - The most crucial aspect of real-time systems
+
+## Intro
+
+Many words have been spoken about RT systems and their supremacy over typical systems in specific fields. But should we trust their name? Real-Time?\
+As usual, the answer is no. RT system just trying to be what they're called, that means, react to triggers like interruptions and response immediately. More often it's just not later than the strictly defined time, called latency.
+
+## Strategy
+
+So how can we measure the 'quality' of our rt system? Let's say we want a task to be weaken up after 200 us. To this task, we can use a timer which will generate interruption when the time runs out and wake up the task.
+Before we put the task to sleep lets remember current time
+
+1. The start point of measuring kernel latencies is when the interrupt occurs.
+2. The First thing kernel must do is observe it and while there is a lot to do and to observe even that part can take some time.
+3. After that, relevant ISR is called to handle interruption(in our case to wake up the task)
+4. Next goes the kernel scheduler whose job is to manage all of the processes working in the system. When our task has been woken up, it was placed on the CPU queue indeed. So now it's the scheduler thing to handle the task ASAP.
+5. When it's time come, CPU starts to process the task and we can end measuring the latency.
+
+After that, we can subtract the current time from the time before we put the task to sleep and get the latency by subtracting time difference with the time given to timer.
+
+The given example is the exact methodology of the [`cyclictest`](http://manpages.ubuntu.com/manpages/cosmic/man8/cyclictest.8.html) test program for testing system latencies which will be using here by us.
+
+### Additional load
+
+To simulate a stressful environment for the system we're testing tools like [`hackbench`](http://manpages.ubuntu.com/manpages/xenial/man8/hackbench.8.html) and [`stressapptest`](http://manpages.ubuntu.com/manpages/trusty/man1/stressapptest.1.html) can be used.
+
+* [`Hackbench`](http://manpages.ubuntu.com/manpages/xenial/man8/hackbench.8.html) - tool for stressing kernel scheduler by creating pairs of threads communicating with each other via sockets
+
+* [`Stressapptest`](http://manpages.ubuntu.com/manpages/trusty/man1/stressapptest.1.html) - program for generating a realistic load of memory, CPU, and I/O by creating a specified amount of threads writing to memory, to file, or communicate with given IP server
+
+## Testing platform
+
+Tests will be performed on an i.mx8 platform with two builds of yocto-linux. A Regular version and a specially configured as RT using rt patch.
```

Comment:
You can link to RP patch sources: https://mirrors.edge.kernel.org/pub/linux/kernel/projects/rt/

Reply:
Link added